monkey/monkey_island/cc/agent_event_handlers/node_update_facade.py

```diff
@@ -0,0 +1,90 @@
+from functools import lru_cache
+from ipaddress import IPv4Address, IPv4Interface
+
+from common.agent_events import AbstractAgentEvent
+from common.types import AgentID, MachineID
+from monkey_island.cc.models import CommunicationType, Machine
+from monkey_island.cc.repository import (
+    IAgentRepository,
+    IMachineRepository,
+    INodeRepository,
+    UnknownRecordError,
+)
+
+
+class NodeUpdateFacade:
+    """
+    A facade to simplify updating nodes and machines based on network-related events.
+    """
+
+    def __init__(
+        self,
+        agent_repository: IAgentRepository,
+        machine_repository: IMachineRepository,
+        node_repository: INodeRepository,
+    ):
+        self._agent_repository = agent_repository
+        self._machine_repository = machine_repository
+        self._node_repository = node_repository
+
+    def get_or_create_target_machine(self, target: IPv4Address):
+        """
+        Gets or creates a target machine from an IP address
+
+        :param target: The IP address representing the target of some event
+        :return: A new or existing machine that matches the target IP address
+        :raises StorageError: If an error occurs while attempting to store a new machine
+        :raises RetrievalError: If an error occurs while attempting to retrieve an existing machine
+        """
+        try:
+            machine_id = self._get_machine_id_by_ip(target)
+            return self._machine_repository.get_machine_by_id(machine_id)
+        except UnknownRecordError:
+            machine = Machine(
+                id=self._machine_repository.get_new_id(),
+                network_interfaces=[IPv4Interface(target)],
+            )
+            self._machine_repository.upsert_machine(machine)
+            return machine
+
+    @lru_cache(maxsize=1024)
+    def _get_machine_id_by_ip(self, ip: IPv4Address) -> MachineID:
+        machines = self._machine_repository.get_machines_by_ip(ip)
+
+        # For now, assume that IPs are unique
+        return machines[0].id
+
+    @lru_cache(maxsize=None)
```

Comment:
Why is there no max size?

Reply:
Because I don't expect the total number of agents to exceed a few thousand and I don't expect them to become irrelevant over the course of a simulation.

Reply:
Does resetting the simulation reset these caches?

Reply:
Good catch!